### PR TITLE
Use SWC to valid client next/navigation hooks usage in server components

### PR DESCRIPTION
**Pull request by huozhi**
### What

Use SWC to check invalid client hooks of `next/navigation` imports in server components. 
Follow up of #62456 

Remove the runtime error APIs for `next/navigation` rsc version.
Add `next/navigation` react-server version alias in turbopack.

This PR also refactored the invalid server layer APIs detection into a map, where key is import path and value is an array of client APIs. During the traversing we will get the import source easily, this makes extending the logic much easier

### Why

Previously we're using the runtime error to check it, but it has to run first then the error will be thrown. If we error first in build time with this check it's much faster and we this align on both side between webpack and turbopack.

### CODE REVIEW

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| buildDuration | 13.9s | 13.9s | N/A |
| buildDurationCached | 8.4s | 6.3s | N/A |
| nodeModulesSize | 198 MB | 198 MB | N/A |
| nextStartRea..uration (ms) | 429ms | 432ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 30.5 kB | 30.5 kB | N/A |
| 3304.HASH.js gzip | 181 B | 181 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.04 kB | 5.04 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 241 B | 242 B | N/A |
| main-HASH.js gzip | 32.2 kB | 32.2 kB | N/A |
| webpack-HASH.js gzip | 1.68 kB | 1.68 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| _app-HASH.js gzip | 196 B | 197 B | N/A |
| _error-HASH.js gzip | 184 B | 184 B | ✓ |
| amp-HASH.js gzip | 505 B | 505 B | ✓ |
| css-HASH.js gzip | 324 B | 325 B | N/A |
| dynamic-HASH.js gzip | 2.5 kB | 2.5 kB | N/A |
| edge-ssr-HASH.js gzip | 258 B | 258 B | ✓ |
| head-HASH.js gzip | 352 B | 352 B | ✓ |
| hooks-HASH.js gzip | 370 B | 371 B | N/A |
| image-HASH.js gzip | 4.21 kB | 4.21 kB | ✓ |
| index-HASH.js gzip | 259 B | 259 B | ✓ |
| link-HASH.js gzip | 2.67 kB | 2.67 kB | N/A |
| routerDirect..HASH.js gzip | 314 B | 312 B | N/A |
| script-HASH.js gzip | 386 B | 386 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 6.57 kB | 6.57 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| _buildManifest.js gzip | 481 B | 484 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| index.html gzip | 529 B | 527 B | N/A |
| link.html gzip | 541 B | 541 B | ✓ |
| withRouter.html gzip | 524 B | 523 B | N/A |
| Overall change | 541 B | 541 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| edge-ssr.js gzip | 95.1 kB | 95.1 kB | N/A |
| page.js gzip | 3.06 kB | 3.07 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 626 B | 623 B | N/A |
| middleware-r..fest.js gzip | 151 B | 151 B | ✓ |
| middleware.js gzip | 25.5 kB | 25.5 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 990 B | 990 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 96.5 kB | 96.5 kB | ✓ |
| app-page-tur..prod.js gzip | 98.3 kB | 98.3 kB | ✓ |
| app-page-tur..prod.js gzip | 92.7 kB | 92.7 kB | ✓ |
| app-page.run...dev.js gzip | 149 kB | 149 kB | ✓ |
| app-page.run..prod.js gzip | 91.2 kB | 91.2 kB | ✓ |
| app-route-ex...dev.js gzip | 21.3 kB | 21.3 kB | ✓ |
| app-route-ex..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 15 kB | 15 kB | ✓ |
| app-route-tu..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| app-route.ru...dev.js gzip | 21 kB | 21 kB | ✓ |
| app-route.ru..prod.js gzip | 14.8 kB | 14.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.8 kB | 9.8 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.52 kB | 9.52 kB | ✓ |
| pages-turbo...prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| pages.runtim...dev.js gzip | 22.9 kB | 22.9 kB | ✓ |
| pages.runtim..prod.js gzip | 22.3 kB | 22.3 kB | ✓ |
| server.runti..prod.js gzip | 50.5 kB | 50.5 kB | ✓ |
| Overall change | 948 kB | 948 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js swc/rsc-check-next-navigation | Change |
| - | - | - | - |
| 0.pack gzip | 1.56 MB | 1.56 MB | ⚠️ +1.3 kB |
| index.pack gzip | 105 kB | 105 kB | N/A |
| Overall change | 1.56 MB | 1.56 MB | ⚠️ +1.3 kB |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 5fb4f69ccc96a25db54bd694c594ed44f112817f

**Review by gnoff — APPROVED** [review]
Happy to see this land. I do think we can make the error message more helpful and left a suggestion in case you want to tackle that here and now vs in a follow up PR

**Review by gnoff — COMMENTED** [review]
Should we also remove `clientHookInServerComponentError` from the client implementation of navigation hooks?

File: test/development/acceptance-app/server-components.test.ts [hunk 1]
```diff
@@ -521,45 +521,39 @@ describe('Error Overlay for server components', () => {
...
-      }, 'success')
-
-      expect(next.cliOutput).toContain(
-        `Error: ${hook} only works in Client Components. Add the "use client" directive at the top of the file to use it. Read more: https://nextjs.org/docs/messages/react-client-hook-in-server-component`
+      expect(await session.hasRedbox()).toBe(true)
+      // In webpack when the message too long it gets truncated with `  | ` with new lines.
+      // So we need to check for the first part of the message.
+      const normalizedSource = await session.getRedboxSource()
+      expect(normalizedSource).toContain(
+        `You're importing a component that needs ${hook}. It only works in a Client Component but none of its parents are marked with "use client"`
```
[the remark this answers is not in the document]
Comment (huozhi) [reply test/development/acceptance-app/server-components.test.ts:553]: Sounds good to improve that as well, we have the similar error message like `Maybe one of these should be marked as a client entry 'use client'` and printing the module trace, mainly for React client hooks before. I can improve this in a separate PR 👍

**Comment by huozhi** [comment]
> Should we also remove `clientHookInServerComponentError` from the client implementation of navigation hooks?

Good catch, deleted it in [183f449](https://github.com/vercel/next.js/pull/63160/commits/183f449fea8a09d007d4a1689d577c86290e0b09)

**Pull request merged**